magento/magento2
# Ensure Quote::collectTotals() correctly handles recursion

### Description
We have written a plug-in for a shipping method. Our plugin requires access the quote object to perform its business logic. We get a reference to the current quote object by calling `Magento\Checkout\Model\Session::getQuote()`. This works correctly and we get a reference to the quote object and can perform the required actions.

Within `Magento\Checkout\Model\Session::getQuote()`, there is a call to `$quote->collectTotals()`. When this runs via our plug-in, Magento ends up calling `$quote->collectTotals()` within an existing call to `$quote->collectTotals()` (recursion). This becomes a problem when we introduce gift cards, which is a feature of Adobe Commerce. Because we are applying the gift cards each time `$quote->collectTotals()` is called, but the second time around the gift cards are already 'used', so they do not get applied to the `$quote`, and therefore disappear from the checkout.


### Manual testing scenarios
Preconditions:
- Adobe Commerce with Gift Card Accounts feature
- Gift card with an available balance less than the total of the cart
- A plug-in within the shipping method totals collection which calls `Magento\Checkout\Model\Session::getQuote()`

Steps to reproduce:
1. Add items to cart to ensure cart total is more than available balance on gift card
1. Apply gift card at cart
1. Proceed to payment step in checkout
1. Notice gift card is applied correctly
1. Select payment method multiple times
1. Notice gift card is (not) applied correctly

### Contribution checklist


### Resolved issues:
1. [x] resolves magento/magento2#37856: Ensure Quote::collectTotals() correctly handles recursion
